src/nix-eval-jobs.cc

```diff
@@ -384,16 +381,72 @@ int main(int argc, char * * argv)
             }
         };
 
-        EvalState initialState(myArgs.searchPath, openStore());
-        Bindings & autoArgs = *myArgs.getAutoArgs(initialState);
+        /* Collect initial attributes to evaluate. This must be done
+           in a separate fork to avoid spawning a download in the
+           parent process. If that happens, worker processes will try
+           to enqueue downloads on their own download threads (which
+           will not exist). */
```

Comment:
Is this basically saying...

> Collect initial attributes to evaluate.

You're a parent thread. You need some stuff computed and downloaded, to figure out how many child threads to spawn and what work to give them.

> [you need to] avoid spawning a download in the parent process. If that happens, worker processes will try to enqueue downloads on their own download threads (which will not exist).

But the act of downloading "mutates your thread DNA". And that causes your children to not work correctly.

> This must be done in a separate fork...

So to avoid polluting your DNA, you create a child and assign it the download. It goes off and downloads the things, and then gives them back to you, and then dies. Which means you have the download artifacts but your DNA remains healthy, so your worker babies will be healthy too.

🙃

Reply:
Yes, exactly!

Reply:
Ah, so it's all processes (`s/thread/process/g`)? Or are there threads too?

Reply:
There are both. The downloader is a thread, `nix-eval-jobs` workers spawn forked processes.

Reply:
Processes don't inherit threads from the parent, just a copy of the memory. So the worker processes would have a static reference to a thread that didn't actually exist.

Reply:
Ah, now I understand. Thanks for explaining!